x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/regression/RSquared.java

```diff
@@ -0,0 +1,151 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.metrics.ExtendedStats;
+import org.elasticsearch.search.aggregations.metrics.ExtendedStatsAggregationBuilder;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregation;
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetricResult;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Calculates R-Squared between two known numerical fields.
+ *
+ * equation: mse = 1 - SSres/SStot
+ * such that,
+ * SSres = Σ(y - y´)^2, The residual sum of squares
+ * SStot =  Σ(y - y_mean)^2, The total sum of squares
+ */
+public class RSquared implements RegressionMetric {
+
+    public static final ParseField NAME = new ParseField("r_squared");
+
+    private static final String PAINLESS_TEMPLATE = "def diff = doc[''{0}''].value - doc[''{1}''].value;return diff * diff;";
+    private static final String SS_RES = "residual_sum_of_squares";
+
+    private static String buildScript(Object...args) {
+        return new MessageFormat(PAINLESS_TEMPLATE, Locale.ROOT).format(args);
+    }
+
+    private static final ObjectParser<RSquared, Void> PARSER =
+        new ObjectParser<>("r_squared", true, RSquared::new);
+
+    public static RSquared fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    public RSquared(StreamInput in) {
+
+    }
+
+    public RSquared() {
+
+    }
+
+    @Override
+    public String getMetricName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public List<AggregationBuilder> aggs(String actualField, String predictedField) {
+        return Arrays.asList(
+            AggregationBuilders.sum(SS_RES).script(new Script(buildScript(actualField, predictedField))),
+            AggregationBuilders.extendedStats(ExtendedStatsAggregationBuilder.NAME + "_actual").field(actualField));
+    }
+
+    @Override
+    public EvaluationMetricResult evaluate(Aggregations aggs) {
+        NumericMetricsAggregation.SingleValue residualSumOfSquares = aggs.get(SS_RES);
+        ExtendedStats extendedStats = aggs.get(ExtendedStatsAggregationBuilder.NAME + "_actual");
+        return residualSumOfSquares == null || extendedStats == null || extendedStats.getCount() == 0 ?
+            null :
+            new Result(1 - (residualSumOfSquares.value() / (extendedStats.getVariance() * extendedStats.getCount())));
```

Comment:
One question, does extendedStats use the maximum likelihood or unbiased estimate of variance? This would be either a `* extendedStats.getCount()` or `* (extendedStats.getCount() - 1)` depending on the answer. I'm guessing this would have been flushed by the tests, but the difference is obviously small for large doc count, I think either way it is probably worth a comment here to say what the aggregation does use.

Reply:
Looking at `InternalExtendedStats` this does use the maximum likelihood form, so multiplying by count is the right thing to do.

Reply:
@tveasey I will add the comment. FWIW, I have tested the calculation on some various datasets to make sure it matched what other tools (namely scikit) calculate for the r_squared metric.